### PR TITLE
Pseudolocalisation

### PR DESCRIPTION
**Pull request by dimaqq**
Pseudolocalisation using Unicode Mathematical Symbols.

Converts:
```po
msgid ""
msgstr ""
"Content-Type: text/plain; charset=utf-8\n"
"Plural-Forms: nplurals = 2; plural = (n !== 1);\n"
"Language: en-PP\n"
"MIME-Version: 1.0\n"
"Content-Transfer-Encoding: 8bit\n"
"Project-Id-Version: \n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: \n"
"PO-Revision-Date: \n"
"Last-Translator: \n"
"Language-Team: \n"

msgid "Username"
msgstr ""

msgid "User ${ match.params.id }"
msgstr ""

#: src/User.js:15
msgid "have ${ x } apple"
msgid_plural "have ${ x } apples"
msgstr[0] ""
msgstr[1] ""
```

Into:
```po
msgid ""
msgstr ""
"Content-Type: text/plain; charset=utf-8\n"
"Plural-Forms: nplurals = 2; plural = (n !== 1);\n"
"Language: en-PP\n"
"MIME-Version: 1.0\n"
"Content-Transfer-Encoding: 8bit\n"
"Project-Id-Version: \n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: \n"
"PO-Revision-Date: \n"
"Last-Translator: \n"
"Language-Team: \n"

msgid "Username"
msgstr "𝓤𝓼𝓮𝓻𝓷𝓪𝓶𝓮"

msgid "User ${ match.params.id }"
msgstr "𝓤𝓼𝓮𝓻 ${match.params.id}"

#: src/User.js:15
msgid "have ${ x } apple"
msgid_plural "have ${ x } apples"
msgstr[0] "𝓱𝓪𝓿𝓮 ${x} 𝓪𝓹𝓹𝓵𝓮"
msgstr[1] "𝓱𝓪𝓿𝓮 ${x} 𝓪𝓹𝓹𝓵𝓮"⏎
```

### CODE REVIEW

**Comment by AlexMost** [comment]
Hi @dimaqq! 
Thanks for the PR. Seems like this feature can be really useful. Will this work for non ASCII characters?

**Comment by dimaqq** [comment]
> Hi @dimaqq!
> Thanks for the PR. Seems like this feature can be really useful. Will this work for non ASCII characters?

There's a technical limitation on what characters can be replaced. Pseudo-localisation in this pull request replaces `[a-zA-Z]` with Mathematical symbols (looks good) and `[0-9]` with Mathematical number symbols (could be better, hard to distinguish from regular numbers).

This means that `naïve` will only have `n,a,v,e` replaced but not `ï`, which is left as is.

I was going after the "common" case of code written in English (`function` and `user_id`) whereby translatable text is typically also in English (for example `ngettext` placeholder in code only supports 2 plural forms while some languages have 3). To be honest I'm aware of but not familiar with coding practice where translatable strings are non-ASCII.

Arguably characters in some other scripts could be replaced with Zalgo; that's not included in the pull request.

**Comment by AlexMost** [comment]
ttag was designed to work with any language in the source files. For example, ngettext can change it's signature with this setting - https://ttag.js.org/docs/ngettext.html#plural-forms-that-are-different-from-default-english. It would be great also to have pseudolocalisation support for any non-ASCII characters. But, anyway I think your PR is a good starting point in this direction and we can mark this command as experimental.

**Comment by dimaqq** [comment]
Great!

P.S. by all means feel free to change up the code in whatever way makes sense, my ts-fu is rather weak :)

**Comment by dimaqq** [comment]
This PR is not enough — the expressions in the pseudo-localised translations are currently broken per https://github.com/ttag-org/ttag/issues/153

Either whitespace around expression needs to be preserved, or that ttag issue addressed.

File: src/commands/pseudo.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        .join("");
+}
+
+function pseudoExpression(msgid: string) {
+    const statement = <ExpressionStatement>tpl("`" + msgid + "`")();
+    const expression = <TemplateLiteral>statement.expression;
+
+    for (const q of expression.quasis) {
+        if (q.value.raw) {
+            q.value.raw = pseudoString(q.value.raw);
```
Comment (dimaqq) [inline src/commands/pseudo.ts]: This is problematic, as it will clobber hex and unicode escape sequences.

File: src/commands/pseudo.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        if (q.value.cooked) {
+            q.value.cooked = pseudoString(q.value.cooked);
```
Comment (dimaqq) [inline src/commands/pseudo.ts]: This should be sufficient? maybe?

**Comment by AlexMost** [comment]
> This PR is not enough — the expressions in the pseudo-localised translations are currently broken per [ttag-org/ttag#153](https://github.com/ttag-org/ttag/issues/153)
> 
> Either whitespace around expression needs to be preserved, or that ttag issue addressed.

That seems like a bug and should be fixed. Going to look at that issue today.

**Pull request merged**